### PR TITLE
Feature/dgr 002 add meeting

### PR DESCRIPTION
**Pull request by AinaArd**

### CODE REVIEW

File: src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import ru.itis.dogger.models.Meeting;
+import ru.itis.dogger.models.Owner;
+
+import java.sql.Timestamp;
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class DetailedMeetingDto {
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java:15]: what is the difference between this class and Meeting model?

File: src/main/java/ru/itis/dogger/dto/MeetingDto.java [hunk 2]
```diff
@@ -1,13 +1,38 @@
...
 import lombok.Data;
+import lombok.NoArgsConstructor;
+import ru.itis.dogger.models.Meeting;
+
 import java.sql.Timestamp;
 
 @Data
+@NoArgsConstructor
+@AllArgsConstructor
 public class MeetingDto {
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/dto/MeetingDto.java:13]: same as DetailedMeetingDto

File: src/main/java/ru/itis/dogger/forms/NewMeetingForm.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package ru.itis.dogger.forms;
+
+import lombok.Data;
+import java.sql.Timestamp;
+
+@Data
+public class NewMeetingForm {
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/forms/NewMeetingForm.java:7]: please, don't use forms. They are equal to dto's, so don't create more classes, just use dto

File: src/main/java/ru/itis/dogger/services/MeetingsService.java [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
 public interface MeetingsService {
-    void addMeeting(MeetingDto dto, Owner creator);
+    void addMeeting(NewMeetingForm form, Owner creator);
+
+    List<Meeting> getAllMeetings();
+
+    Optional<Meeting> getMeetingById(long id);
 
-    String getAllMeetings();
+    void joinMeeting(Owner currentUser, long meetingId);
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/services/MeetingsService.java:18]: change long to Long

File: src/main/java/ru/itis/dogger/services/MeetingsService.java [hunk 5]
```diff
@@ -1,11 +1,19 @@
...
+import ru.itis.dogger.models.Meeting;
 import ru.itis.dogger.models.Owner;
 
+import java.util.List;
+import java.util.Optional;
+
 
 public interface MeetingsService {
-    void addMeeting(MeetingDto dto, Owner creator);
+    void addMeeting(NewMeetingForm form, Owner creator);
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/services/MeetingsService.java:12]: use dto

File: src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java [hunk 6]
```diff
@@ -21,22 +22,38 @@ public MeetingsServiceImpl(MeetingsRepository meetingsRepository) {
...
     @Override
-    public String getAllMeetings() {
-        List<Meeting> meetings = meetingsRepository.findAll();
-        return new Gson().toJson(meetings);
+    public List<Meeting> getAllMeetings() {
+        return meetingsRepository.findAll();
+    }
+
+    @Override
+    public Optional<Meeting> getMeetingById(long id) {
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java:45]: long -> Long

File: src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java [hunk 7]
```diff
@@ -21,22 +22,38 @@ public MeetingsServiceImpl(MeetingsRepository meetingsRepository) {
...
+        return meetingsRepository.findById(id);
+    }
+
+    @Override
+    public void joinMeeting(Owner currentUser, long meetingId) {
+        boolean isAlreadyJoined = currentUser.getMeetings().stream()
+                .anyMatch(m -> m.getId().equals(meetingId));
+        if (!isAlreadyJoined) {
+            Meeting meeting = meetingsRepository.findById(meetingId).orElseThrow(IllegalArgumentException::new);
```
Comment (AinaArd) [inline src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java:54]: firstly -> return Optional
secondly -> don't use orElseThrow. You should handle case, when meeting is null

File: src/main/resources/requests.http [hunk 8]
```diff
@@ -26,6 +26,20 @@ Content-Type: application/json
 
 ###
 
+POST http://localhost:8080/signUp
```
Comment (AinaArd) [inline src/main/resources/requests.http:29]: no need to push this file

File: src/main/resources/requests.http [hunk 9]
```diff
@@ -43,11 +57,35 @@ Content-Type: application/json
...
 
+
+
 <> 2020-03-31T080522.200.json
 <> 2020-03-31T080024.200.json
 
 ###
 
+POST http://localhost:8080/login
+Accept: */*
```
Comment (AinaArd) [inline src/main/resources/requests.http:68]: it's better to update api docs

File: src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java (hunk 1 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java:15]: Meeting model has ManyToMany relationship with Owner model. For now I don't see another decision for escaping StackOverFlowError occuring in response for GET /meetings/{id}. It depends on a huge amount of annotations related to other functional (@JsonIngore and etc), we cannot just delete them and not harm other functional.
I believe that we can use only Meeting model for purpose of GET /meetings/{id} but I can't spend more time on checking and maybe changing whole structure including other classes to resolve StackOverFlow problem.

File: src/main/java/ru/itis/dogger/dto/MeetingDto.java (hunk 2 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/dto/MeetingDto.java:13]: They are not the same. MeetingDto excludes detailed info about creator and his meetings, and info about participants and their detailed info. For Mettings main page this info is not requested.

File: src/main/java/ru/itis/dogger/services/MeetingsService.java (hunk 5 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/services/MeetingsService.java:12]: I just renamed Dto to Form to separate new meeting creation info model and transfer meeting info model

File: src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java (hunk 1 above)
Comment (AinaArd) [reply src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java:15]: decision - make a hashmap and put there only needed things. Then no need of dto's

Comment (emilg1101) [reply src/main/java/ru/itis/dogger/dto/DetailedMeetingDto.java:18]: > decision - make a hashmap and put there only needed things. Then no need of dto's

image

File: src/main/java/ru/itis/dogger/controllers/MeetingsController.java [hunk 10]
```diff
@@ -22,17 +28,37 @@ public MeetingsController(MeetingsService meetingsService) {
...
     @PreAuthorize("isAuthenticated()")
     public ResponseEntity<?> getMeetingsList(@RequestHeader(name = "Authorization") String token, Authentication authentication) {
-        return ResponseEntity.ok(meetingsService.getAllMeetings());
+        List<MeetingDto> meetingDtos = meetingsService.getAllMeetings().stream().map(mtg -> MeetingDto.from(mtg)).collect(Collectors.toList());
+        return ResponseEntity.ok(meetingDtos);
+    }
+
+    @GetMapping("/meetings/{meetingId}")
+    @PreAuthorize("isAuthenticated()")
+    public ResponseEntity<?> getDetailedMeeting(@RequestHeader(name = "Authorization") String token, @PathVariable long meetingId) {
```
[the remark this answers is not in the document]
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/controllers/MeetingsController.java]: done

File: src/main/java/ru/itis/dogger/controllers/MeetingsController.java [hunk 11]
```diff
@@ -22,17 +28,37 @@ public MeetingsController(MeetingsService meetingsService) {
...
+        Meeting meeting = meetingsService.getMeetingById(meetingId).orElse(null);
```
[the remark this answers is not in the document]
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/controllers/MeetingsController.java]: fixed

File: src/main/java/ru/itis/dogger/controllers/MeetingsController.java [hunk 12]
```diff
@@ -22,17 +28,37 @@ public MeetingsController(MeetingsService meetingsService) {
...
+        if (meeting == null) {
+            return ResponseEntity.notFound().build();
+        } else
+            return ResponseEntity.ok(DetailedMeetingDto.from(meeting));
+    }
+
+    @PostMapping("/meetings/{meetingId}")
+    @PreAuthorize("isAuthenticated()")
+    public ResponseEntity<?> joinMeeting(@RequestHeader(name = "Authorization") String token,
+                                         @PathVariable long meetingId, Authentication authentication) {
```
[the remark this answers is not in the document]
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/controllers/MeetingsController.java]: done

File: src/main/java/ru/itis/dogger/forms/NewMeetingForm.java (hunk 3 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/forms/NewMeetingForm.java]: changed

File: src/main/java/ru/itis/dogger/services/MeetingsService.java (hunk 4 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/services/MeetingsService.java]: done

File: src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java (hunk 6 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java]: done

File: src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java (hunk 7 above)
Comment (kamilamie) [reply src/main/java/ru/itis/dogger/services/MeetingsServiceImpl.java]: fixed

File: src/main/resources/requests.http (hunk 8 above)
Comment (kamilamie) [reply src/main/resources/requests.http]: если внесешь изменения, которые возможно повлияют на митинги, лучше иметь под рукой готовые запросы, чтобы протестить (надо бы тесты написать конечно, но мне лень)

File: src/main/resources/requests.http (hunk 9 above)
Comment (kamilamie) [reply src/main/resources/requests.http]: обновила

**Pull request merged**